tests/test_omegaconf.py

```diff
@@ -518,3 +518,128 @@ def test_resolve(cfg: Any, expected: Any) -> None:
 def test_resolve_invalid_input() -> None:
     with raises(ValueError):
         OmegaConf.resolve("aaa")  # type: ignore
+
+
+@mark.parametrize(
+    "cfg, expected",
+    [
+        # dict:
+        (
+            DictConfig({
+                "a": 10,
+                "b": {
+                    "c": "???",
+                    "d": "..."
+                }
+            }),
+            {"b.c"}
+        ),
+        (
+            DictConfig({
+                "a": "???",
+                "b": {
+                    "foo": "bar",
+                    "bar": "???",
+                    "more": {
+                        "missing": "???",
+                        "available": "yes"
+                    }
+                },
+            }),
+            {"a", "b.bar", "b.more.missing"}
+        ),
+        (
+            DictConfig({
+                "a": "a",
+                "b": {
+                    "foo": "bar",
+                    "bar": "foo",
+                },
+            }),
+            set()
+        ),
+        (
+            {
+                "foo": "bar",
+                "bar": "???",
+                "more": {
+                    "foo": "???",
+                    "bar": "foo"
+                }
+            },
+            {"bar", "more.foo"}
+        ),
+
+        # list:
+        (
+            ListConfig([
+                "???",
+                "foo",
+                "bar",
+                "???",
+                77
+            ]),
+            {"0", "3"}
+        ),
+        (
+            ListConfig([
+                "",
+                "foo",
+                "bar"
+            ]),
+            set()
+        ),
+        (
+            ["foo", "bar", "???"],
+            {"2"}
+        ),
+
+        # mixing:
+        (
+            ListConfig([
+                "???",
+                "foo",
+                DictConfig({
+                    "a": True,
+                    "b": "???",
+                    "c": ["???", None],
+                    "d": {
+                        "e": "???",
+                        "f": "fff",
+                        "g": [True, "???"]
+                    }
+                }),
+                "???",
+                77
+            ]),
+            {"0", "2.b", "2.c.0", "2.d.e", "2.d.g.1", "3"}
+        ),
+        (
+            {
+                "list": [
+                    0,
+                    DictConfig({
+                        "foo": "???",
+                        "bar": None
+                    }),
+                    "???",
+                    ["???", 3, False]
+                ],
+                "x": "y",
+                "y": "???"
+            },
+            {"list.1.foo", "list.2", "list.3.0", "y"}
```

Comment:
for lists, this would be a better syntax:
```suggestion
            {"list[1].foo", "list[2]", "list[3][0]", "y"}
```

Reply:
@omry What about a plain list like [1, ???, 2] - do you think it's better like `{[1]}`, or  `{1}`?

Reply:
Yes, I think it's okay to always use brackets for lists.
In fact, you can use `node._get_full_key(key)` on nodes to get the proper representation.
try it as an alternative to constructing the key yourself.

Reply:
Great I dropped constructing the prefix on my own, using _get_full_key instead.

Reply:
The `_get_full_key` method gets the key relative to the root node.
```python
>>> cfg = OmegaConf.create({"a": {"b": {"c": "???"}}})
>>> subcfg = cfg.a
>>> print(subcfg)
{'b': {'c': '???'}}
>>> OmegaConf.missing_keys(subcfg)
{'a.b.c'}
```
I would have expected instead that the missing keys are relative to the config that is passed to `OmegaConf.missing_keys`:
```python
>>> OmegaConf.missing_keys(subcfg)
{'b.c'}  # This is what I would expect
```

Reply:
If you both agree with me about the expected behavior, there are a few routes to implementation.
I think the easiest would be to temporarily set the `_parent` and the `_metadata.key` for the `cfg`, so that `_get_full_key` will stop moving up the tree at that point.
```python
    @staticmethod
    def missing_keys(cfg: Any) -> Set[str]:

        ...  # same as before

        old_parent = cfg.__dict__["_parent"]
        old_key = cfg._metadata.key
        try:
            cfg.__dict__["_parent"] = None
            cfg._metadata.key = None
            gather(cfg)
            return missings
        finally:
            cfg.__dict__["_parent"] = old_parent
            cfg._metadata.key = old_key
```

Reply:
Good point about the config root.
Yes, I think it would ideally behave as you are describing.
A solution can either be what you are proposing or allowing `_get_full_key` to take an optional root (if not provided it can use the real root, otherwise it can use the provided root).

When it hits the root (instead of when there is no parent as is the case now) it can stop building the key.

This is more complicated, but also cleaner because it's thread safe. Currently OmegaConf is not thread safe so it can work, but in general it's better form to not mutate an object when you are inspecting it, not even temporarily.

Thoughts?

Reply:
Yes, passing a `root` to `_get_full_key` is appealing, and the functionality might be useful elsewhere in the library. Definitely harder to implement though :)

> Currently OmegaConf is not thread safe so it can work, but in general it's better form to not mutate an object when you are inspecting it, not even temporarily.

Ok, interesting. I think it will be thread-safe (though less efficient) if we use `deepcopy`:
```
    @staticmethod
    def missing_keys(cfg: Any) -> Set[str]:

        ...  # same as before

        cfg = deepcopy(cfg)
        cfg.__dict__["_parent"] = None
        cfg._metadata.key = None
        gather(cfg)
        return missings
```


> 3. Even if they do call it on a config node, it's not strictly wrong to produce the full keys.

Fair enough.

Reply:
In general, the use case for this function is to determine the missing keys in the config for the purposes of producing an error. This will almost certainly be performed on the entire config object and not on a subnode.

Reply:
Yup, makes sense. Ok, resolving this as low priority.